libctru/source/services/dsp.c

```diff
@@ -0,0 +1,254 @@
+	}
+
+	return 0;
+}
+
+
+Result DSP_GetHeadphoneStatus(bool* is_inserted)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x001F,0,0);
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	*is_inserted = cmdbuf[2] & 0xFF;
+	return cmdbuf[1];
+}
+
+
+Result DSP_FlushDataCache(u32 address, u32 size)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x13,2,2);
+	cmdbuf[1] = address;
+	cmdbuf[2] = size;
+	cmdbuf[3] = IPC_Desc_SharedHandles(1);
+	cmdbuf[4] = CUR_PROCESS_HANDLE;
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
+
+
+Result DSP_InvalidateDataCache(u32 address, u32 size)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x14,2,2);
+	cmdbuf[1] = address;
+	cmdbuf[2] = size;
+	cmdbuf[3] = IPC_Desc_SharedHandles(1);
+	cmdbuf[4] = CUR_PROCESS_HANDLE;
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
+
+
+
+Result DSP_SetSemaphore(u16 value)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x7,1,0);
+	cmdbuf[1] = value;
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
+
+
+
+Result DSP_SetSemaphoreMask(u16 mask)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x17,1,0);
+	cmdbuf[1] = mask;
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
+
+Result DSP_GetSemaphoreHandle(Handle* semaphore)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x16,0,0);
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	*semaphore = cmdbuf[3];
+	return cmdbuf[1];
+}
+
+Result DSP_LoadComponent(u8 const* component,u32 size,u16 prog_mask,u16 data_mask,bool * is_loaded)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x11,3,2);
+	cmdbuf[1] = size;
+	cmdbuf[2] = prog_mask;
+	cmdbuf[3] = data_mask;
+	cmdbuf[4] = IPC_Desc_Buffer(size,IPC_BUFFER_R);
+	cmdbuf[5] = (u32) component;
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	*is_loaded = cmdbuf[2] & 0xFF;
+	return cmdbuf[1];
+}
+
+
+
+Result DSP_UnloadComponent(void)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x12,0,0);
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
+
+Result DSP_RegisterInterruptEvents(Handle handle, u32 interrupt, u32 channel)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x15,2,2);
+	cmdbuf[1] = interrupt;
+	cmdbuf[2] = channel;
+	cmdbuf[3] = IPC_Desc_SharedHandles(1);
+	cmdbuf[4] = handle;
+	if ((ret = svcSendSyncRequest(dspHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
+
+
+Result DSP_ReadPipeIfPossibleEx(u32 channel,u32 unk1, u8 const *buffer, u16 length, u16* length_read)
```

Comment:
I didn't expose it because from what I saw, unk1 is always set to 0.
The plan was to have DSP_ReadPipeIfPossible signature change and replace DSP_ReadPipeIfPossibleEx if we ever find use for this unknown parameter